src/Raven.Server/Documents/ResourceCache.cs

```diff
@@ -228,9 +271,15 @@ public IDisposable RemoveLockAndReturn(string databaseName, Action<TResource> on
                 {
                     _caseInsensitive.TryUpdate(databaseName, resourceLocked, current);
                     _resourceDetails.TryRemove(current, out _);
-                    RemoveCaseSensitive(databaseName, current);
+                    RemoveCaseSensitive(databaseName);
+
+                    // We need to refresh all of them.
+                    _readonlyCaseSensitive = _caseSensitive.ToFrozenDictionary();
+                    _readonlyCaseInsensitive = _caseInsensitive.ToFrozenDictionary();
+                    _readonlyResourceDetails = _resourceDetails.ToFrozenDictionary();
```

Comment:
That opens us up to race conditions, no? Shouldn't we do this under the lock (with the understanding that it can be expensive)?

Reply:
This is still under the lock so it's fine

Reply:
It does, and it already existed in the current version. I will try to diminish the exposure a bit more. https://github.com/ravendb/ravendb/pull/17942#discussion_r1441767696